open_spiel/games/dou_dizhu/dou_dizhu_utils.cc

```diff
@@ -0,0 +1,860 @@
+    if(counter++ == kicker_steps) break;
+  }
+
+  hand[kicker_rank] = (params.kicker_type == kSolo? 1: 2);
+  return hand;
+}
+
+
+int SingleTrioCombHandToActionId(const std::array<int, kNumRanks>& hand){
+
+  int trio_rank, kicker_rank;
+  int trio_counter = 0, kicker_counter = 0;
+  for(int rank = 0; rank < kNumRanks; ++rank){
+    if(hand[rank] == 3) {trio_counter++; trio_rank = rank;}
+    else if(hand[rank] == 1 || hand[rank] == 2) {kicker_counter++; kicker_rank=rank;}
+    else if(hand[rank] == 4) SpielFatalError("There cannot be a bomb");
+  }
+  SPIEL_CHECK_EQ(trio_counter, 1);
+  SPIEL_CHECK_EQ(kicker_counter, 1);
+  
+  int action;
+  if(hand[kicker_rank] == 1) action = kTrioWithSoloActionBase;
+  else action = kTrioWithPairActionBase;
+  // one of the rank had already been taken by the trio
+  if(hand[kicker_rank] == 1) action += trio_rank * (kNumRanks - 1);
+  else action += trio_rank * (kNumRanks - 3);  // the jokers cannot be the pair
+  int kicker_steps = 0;
+  for(int rank = 0; rank < kNumRanks; ++rank){
+    if(rank == trio_rank) continue;
+    if(rank == kicker_rank) break;
+    kicker_steps++;
+  }
+  action += kicker_steps;
+  return action;
+}
+
+
+
+void SearchSingleTrioCombActions(std::vector<Action>* actions, const std::array<int, kNumRanks>& hand, int prev_action = kInvalidAction){
+  TrioCombParams prev_action_params;
+  int start_rank;
+  if(prev_action == kInvalidAction) start_rank = 0;
+  else{
+    prev_action_params = GetSingleTrioCombParams(prev_action);
+    start_rank = prev_action_params.chain_head + 1;
+  }
+  // enumerate possible trio
+  for(int rank = start_rank; rank < kNumRanks - 2; ++rank){
+    if(hand[rank] < 3) continue;
+    for(int kicker = 0; kicker < kNumRanks; ++kicker){
+      if(!hand[kicker] || kicker == rank) continue;
+      std::vector<KickerType> all_kicker_types;
+      if(prev_action != kInvalidAction){
+        if(hand[kicker] >= prev_action_params.kicker_type)
+         all_kicker_types.push_back(prev_action_params.kicker_type);
+      }else{
+        for(int i = 1; i <= std::min(hand[kicker], 2); ++i) 
+          all_kicker_types.push_back(static_cast<KickerType>(i));
+      }
+      for(auto n: all_kicker_types){
+        std::array<int, kNumRanks> used_hand{};
+        used_hand[rank] = 3;
+        used_hand[kicker] = static_cast<int>(n);
+        actions->push_back(SingleTrioCombHandToActionId(used_hand));
+      }
+    }
+  }
+}
+
+
+
+
+
+// a dfs backtrack algorithm to compute action ids / hands for airplane combinations
+// if target_count = -1, then the goal of this algorithm is to find the kicker_id
+// of ans_hand, stored in count reference
+// otherwise, the goal is to find a hand whose kicker_id is target_count
+// and the result hand is stored in ans_hand reference
+bool dfs_airplane_kicker(int chain_length, int depth,
+              int target_count, int& count, int max_search_rank,
+      int* used_rank, int* ans_hand, 
```

Comment:
I think I was unclear/did a bad job reviewing this. 

Originally, these were `std::array<int, kNumRanks>&`. I think that the type here should be `absl::Span<int>`. 

You would convert them into a span like this:

```
 dfs_airplane_kicker(chain_length, depth+1, target_count, count, rank, absl::MakeSpan(used_rank), absl::MakeSpan(ans_hand), kicker_type)
```